src/behaviors/focusTrap.ts

```diff
@@ -0,0 +1,161 @@
+import {iterateTabbableElements} from '../utils/iterateTabbable'
+import {polyfill as eventListenerSignalPolyfill} from '../polyfills/eventListenerSignal'
+
+eventListenerSignalPolyfill()
+
+interface FocusTrapMetadata {
+  container: HTMLElement
+  controller: AbortController
+  initialFocus?: HTMLElement
+  originalSignal: AbortSignal
+}
+
+const suspendedTrapStack: FocusTrapMetadata[] = []
+let activeTrap: FocusTrapMetadata | undefined = undefined
+
+function tryReactivate() {
+  const trapToReactivate = suspendedTrapStack.pop()
+  if (trapToReactivate) {
+    focusTrap(trapToReactivate.container, trapToReactivate.initialFocus, trapToReactivate.originalSignal)
+  }
+}
+
+// @todo If AbortController.prototype.follow is ever implemented, that
+// could replace this function. @see https://github.com/whatwg/dom/issues/920
+function followSignal(signal: AbortSignal): AbortController {
+  const controller = new AbortController()
+  signal.addEventListener('abort', () => {
+    controller.abort()
+  })
+  return controller
+}
+
+/**
+ * Returns the first focusable child of `container`. If `lastChild` is true,
+ * returns the last focusable child of `container`.
+ * @param container
+ * @param lastChild
+ */
+function getFocusableChild(container: HTMLElement, lastChild = false) {
+  return iterateTabbableElements(container, {reverse: lastChild, strict: true}).next().value
+}
+
+/**
+ * Traps focus within the given container.
+ * @param container The container in which to trap focus
+ * @returns AbortController - call `.abort()` to disable the focus trap
+ */
+export function focusTrap(container: HTMLElement, initialFocus?: HTMLElement): AbortController
+
+/**
+ * Traps focus within the given container.
+ * @param container The container in which to trap focus
+ * @param abortSignal An AbortSignal to control the focus trap.
+ */
+export function focusTrap(container: HTMLElement, initialFocus: HTMLElement | undefined, abortSignal: AbortSignal): void
+export function focusTrap(
+  container: HTMLElement,
+  initialFocus?: HTMLElement,
+  abortSignal?: AbortSignal
+): AbortController | void {
+  // Set up an abort controller if a signal was not passed in
+  const controller = new AbortController()
+  const signal = abortSignal ?? controller.signal
+
+  container.setAttribute('data-focus-trap', 'active')
+  let lastFocusedChild: HTMLElement | undefined = undefined
+  const firstFocusableChild = getFocusableChild(container)
+  const lastFocusableChild = getFocusableChild(container, true)
```

Comment:
This shouldn't be a problem with the Hook since we will re-run the hook whenever `containerRef` changes, but do we need to worry about containers that have dynamic content in them? (DM me for an example in an unreleased GitHub project.)

Reply:
If you decide to do this in a follow up can you create an issue for it just in case we forget? This behavior will need to be supported in Memex